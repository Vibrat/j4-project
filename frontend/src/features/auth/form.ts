import z from "zod"

export const authSchema = z.object({
  email: z.string().email({ message: "Invalid email" }),
  password: z.string().min(1, { message: "Password is required" }),
})

export type AuthSchema = z.infer<typeof authSchema>