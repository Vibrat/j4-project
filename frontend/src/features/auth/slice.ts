import { createSlice } from "@reduxjs/toolkit";
import type { PayloadAction } from "@reduxjs/toolkit";

export interface AuthState {
  authenticated: boolean;
  email?: string;
  token?: string;
}

export const intialAuthState: AuthState = {
  authenticated: false,
};

export const authSlice = createSlice({
  name: "auth",
  initialState: intialAuthState,
  reducers: {
    login: (state: AuthState, action: PayloadAction<AuthState>) => {
      const payload = action.payload
      state = {
        ...state,
        ...payload
      }
      return state
    },
    logout: (state: AuthState, action: PayloadAction<void>) => {
      return {
        ...intialAuthState
      }
    }
  }
})

export const selectAuth = (state: any) => state.auth
export const { login, logout } = authSlice.actions
export default authSlice.reducer