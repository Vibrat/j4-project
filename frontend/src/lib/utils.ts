import { type ClassValue, clsx } from "clsx"
import { twMerge } from "tailwind-merge"

export function cn(...inputs: ClassValue[]) {
  return twMerge(clsx(inputs))
}

export function isLocalStorageAvailable() {
  return typeof window !== "undefined" && window.localStorage;
}