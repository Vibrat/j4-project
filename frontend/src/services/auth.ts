import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/query/react'
import { API_BASE_URL } from './config'

export type Credential = {
  email: string
  password: string
}

// Define a service using a base URL and expected endpoints
export const authAPI = createApi({
  reducerPath: "authAPI",
  baseQuery: fetchBaseQuery({ 
    baseUrl: `${API_BASE_URL}/auth`,
  }),
  endpoints: (builder) => ({
    loginServer: builder.mutation({
      query: (cred: Credential) => ({
        url: 'login',
        method: 'POST',
        body: cred
      }),
    }),
  }),
})

// Export hooks for usage in functional components, which are
// auto-generated based on the defined endpoints
export const { useLoginServerMutation } = authAPI