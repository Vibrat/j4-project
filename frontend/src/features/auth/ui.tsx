"use client";
import { zodResolver } from "@hookform/resolvers/zod";
import React from "react";
import { useForm } from "react-hook-form";
import { AuthSchema, authSchema } from "@/features/auth/form";
import {
  Form,
  FormControl,
  FormField,
  FormItem,
  FormMessage,
} from "@/components/ui/form";
import { Input } from "@/components/ui/input";
import { Button } from "@/components/ui/button";
import { useDispatch } from "react-redux";
import { login } from "./slice";
import { useLoginServerMutation } from "@/services/auth";
import { useToast } from "@/components/ui/use-toast";
import MasterSocker from "@/services/notification";

export type AuthFormProps = {
  classNames?: string;
  onSubmit?: (values: AuthSchema) => void;
};

const AuthForm: React.FC<AuthFormProps> = ({ classNames, onSubmit: extraSubmit }) => {
  const [loginServer, { isLoading, isSuccess, isError }] = useLoginServerMutation();
  const dispatch = useDispatch();
  const { toast } = useToast()
  const form = useForm<AuthSchema>({
    resolver: zodResolver(authSchema),
    defaultValues: {
      email: "",
      password: "",
    },
  });

  function onSubmit(values: AuthSchema) {
    loginServer(values)
      .then((response) => {
        if (!response.error) {
          dispatch(
            login({
              authenticated: true,
              email: values.email,
              token: response.data.token,
            })
          )
          extraSubmit?.(response.data);
          MasterSocker?.connect()
        } else {
          // toast an unauthenticated case here
          toast({
            title: "Unauthenticated",
            description: "Please try again",
            variant: "destructive",
          })
        }
      })
      .catch((error) => {
        console.error(error);
      });
  }
  
  return (
    <Form {...form}>
      <form onSubmit={form.handleSubmit(onSubmit)} className={classNames}>
        <FormField
          control={form.control}
          name="email"
          render={({ field }) => (
            <FormItem>
              <FormControl>
                <Input {...field} placeholder="email" />
              </FormControl>
              <FormMessage />
            </FormItem>
          )}
        />
        <FormField
          control={form.control}
          name="password"
          render={({ field }) => (
            <FormItem>
              <FormControl>
                <Input {...field} type="password" placeholder="password" />
              </FormControl>
              <FormMessage />
            </FormItem>
          )}
        />
        <Button type="submit">Login / Register</Button>
      </form>
    </Form>
  );
};

export default AuthForm;
