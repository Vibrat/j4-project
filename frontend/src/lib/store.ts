import { configureStore } from '@reduxjs/toolkit'
import authReducer from '@/features/auth/slice'
import localStorageMiddleware from '@/features/auth/middleware'
import { authAPI } from '@/services/auth'
import { videoAPI } from '@/services/video'
import { notificationAPI } from '@/services/notification'

export const appStore = () => {
  return configureStore({
    reducer: {
      auth: authReducer,
      [videoAPI.reducerPath]: videoAPI.reducer,
      [authAPI.reducerPath]: authAPI.reducer,
      [notificationAPI.reducerPath]: notificationAPI.reducer
    },
    middleware: (getDefaultMiddleware) => getDefaultMiddleware()
      .prepend(localStorageMiddleware.middleware)
      .concat(authAPI.middleware)
      .concat(videoAPI.middleware)
      .concat(notificationAPI.middleware)
  })
}

// Infer the type of makeStore
export type AppStore = ReturnType<typeof appStore>
// Infer the `RootState` and `AppDispatch` types from the store itself
export type RootState = ReturnType<AppStore['getState']>
export type AppDispatch = AppStore['dispatch']