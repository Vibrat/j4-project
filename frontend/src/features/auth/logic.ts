import { isLocalStorageAvailable } from "@/lib/utils";
import { intialAuthState } from "./slice";

export function validateAuth() {
  const auth = isLocalStorageAvailable() ? localStorage.getItem("auth") : JSON.stringify(intialAuthState);
  if (!auth) return intialAuthState;
  return auth ? JSON.parse(auth) : intialAuthState;
}