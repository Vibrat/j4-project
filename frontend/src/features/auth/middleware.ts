import { createListenerMiddleware, isAnyOf } from "@reduxjs/toolkit"
import { login, logout } from "./slice"
import { isLocalStorageAvailable } from "@/lib/utils"

const localStorageMiddleware = createListenerMiddleware()
localStorageMiddleware.startListening({
  matcher: isAnyOf(login, logout),
  effect: async (action, listenerApi) => {
    switch (action.type) {
      case "auth/login":
        isLocalStorageAvailable() && localStorage.setItem('auth', JSON.stringify(action.payload))
        break
      case "auth/logout":
        isLocalStorageAvailable() && localStorage.removeItem('auth')
    }
  }
})

export default localStorageMiddleware