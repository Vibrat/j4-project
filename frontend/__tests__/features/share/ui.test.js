import { render, fireEvent, act } from "@testing-library/react";
import ShareForm from "@/features/share/ui";
import StoreProvider from "@/app/storeProvider";
import { http, HttpResponse } from "msw";
import { setupServer } from "msw/node";
import { API_BASE_URL } from "@/services/config";
import { appStore } from "@/lib/store";
import { login } from "@/features/auth/slice";

jest.mock("next/navigation", () => ({
  useRouter: () => jest.fn(),
}));

export const handlers = [
  http.post(`${API_BASE_URL}/videos`, async () => {
    return HttpResponse.json({});
  }),
];

const server = setupServer(...handlers);

describe("ShareUI", () => {
  beforeAll(() => server.listen());
  afterEach(() => server.resetHandlers());
  afterAll(() => server.close());

  it("should render correctly", () => {
    const { container } = render(
      <StoreProvider>
        <ShareForm />
      </StoreProvider>
    );
    expect(container).toMatchSnapshot();
  });

  it("should not trigger onSubmit when unauthenticated", async () => {
    const afterSubmitCallBack = jest.fn();
    const { getByRole, getByPlaceholderText } = render(
      <StoreProvider>
        <ShareForm onSubmitCallBack={afterSubmitCallBack} />
      </StoreProvider>
    );
    const submitButton = getByRole("button");
    const youtubeField = getByPlaceholderText("https://youtube.com/xsyqgsa");
    await act(() => {
      fireEvent.change(youtubeField, {
        target: { value: "https://youtube.com/abcd" },
      });
      fireEvent.click(submitButton);
    });
    expect(afterSubmitCallBack).toBeCalledTimes(0);
  });

  it("should trigger onSubmit when authenticated", async () => {
    const store = appStore();
    const afterSubmitCallBack = jest.fn();
    
    // perform login
    store.dispatch(login({ authenticated: true, email: "test@gmail.come",token: "ABC",}));
    
    // build component
    const { getByRole, getByPlaceholderText } = render(
      <StoreProvider store={store}>
        <ShareForm onSubmitCallBack={afterSubmitCallBack} />
      </StoreProvider>
    );

    // fire submission
    const submitButton = getByRole("button");
    const youtubeField = getByPlaceholderText("https://youtube.com/xsyqgsa");
    await act( async () => {
      fireEvent.change(youtubeField, {
        target: { value: "https://youtube.com/abcd" },
      });
      fireEvent.click(submitButton);
    });

    // test if it's called only one
    expect(afterSubmitCallBack).toBeCalledTimes(1);
  });
});
