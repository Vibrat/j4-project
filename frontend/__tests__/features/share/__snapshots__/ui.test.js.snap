// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShareUI should render correctly 1`] = `
<div>
  <form>
    <div
      class="space-y-2"
    >
      <input
        aria-describedby=":r0:-form-item-description"
        aria-invalid="false"
        class="flex h-9 w-full rounded-md border border-input bg-transparent px-3 py-1 text-sm shadow-sm transition-colors file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:cursor-not-allowed disabled:opacity-50"
        id=":r0:-form-item"
        name="youtubeURL"
        placeholder="https://youtube.com/xsyqgsa"
        value=""
      />
    </div>
    <button
      class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 bg-primary text-primary-foreground shadow hover:bg-primary/90 h-9 px-4 py-2 w-full"
      type="submit"
    >
      Share
    </button>
  </form>
</div>
`;
